### ISSUE TITLE
How to finetune LLaVA with your data

### ISSUE THREAD

**Issue by zhuzilin** [HYPJUDY/Sparkles#2]
Thank you for your great work on allowing multiple images in multimodal model!

In your paper, you mentioned about finetuning LLaVA v1.5 with SparklesDialog. Could you share how to reoganize the input of LLaVA for multiple images? As the origin LLaVA implementation requires to move the image to the very front. Thank you!

**Comment by HYPJUDY** [comment]
Thank you for your interest in our work! 

The original LLaVA architecture does indeed support inputting multiple images, as detailed in the code snippet from [llava/model/llava_arch.py#L99-L240](https://github.com/haotian-liu/LLaVA/blob/2120e82f9209094dbe5eba1414b763c31edd8aa8/llava/model/llava_arch.py#L99-L240). The `preprocess_multimodal()` function, typically used for single-image inputs, concatenates this image with a sentence by moving the image to the very front (as detailed at [preprocess_multimodal()](https://github.com/haotian-liu/LLaVA/blob/2120e82f9209094dbe5eba1414b763c31edd8aa8/llava/train/train.py#L302C23-L323)).

To extend LLaVA's functionality for multiple images, we modified the data construction process. Here's how we adapt the [`__getitem__` method](https://github.com/haotian-liu/LLaVA/blob/2120e82f9209094dbe5eba1414b763c31edd8aa8/llava/train/train.py#L660) of LLaVA's dataset class to handle an array of images:


```python
def __getitem__(self, i) -> Dict[str, torch.Tensor]:

    # ... existing code ...

    # Handling single image case
    if 'image' in sources[0]:
        # ... original single image processing code ...
    
+   # New block: Handling multiple images case
+   elif "images" in sources[0]:
+       # Process each image in the 'images' list
+       for image_item in self.list_data_dict[i]['images']:
+           # ... existing image loading and processing code ...
+           image_item['image'] = image  # Attach processed image to the image_item
+
+       # Attention: skip preprocess_multimodal to avoid moving the image to the very front
+       # sources = preprocess_multimodal(copy.deepcopy([e["conversations"] for e in sources]), self.data_args)
+       sources = copy.deepcopy([e["conversations"] for e in sources])

    # Handling the case where there are no images in the data
    else:
        sources = copy.deepcopy([e["conversations"] for e in sources])

    # ... further processing ...

    # Compiling all the data together, now accommodating multiple images as well
    if 'image' in self.list_data_dict[i]:
        data_dict['images'] = image  # single image case
+   elif 'images' in self.list_data_dict[i]:
+       # Create a list of all processed images for the multiple images case
+       data_dict['images'] = [image_item["image"] for image_item in self.list_data_dict[i]['images']]

    # ... existing code for multimodal check ...

    return data_dict
```

In summary, these changes enable handling multiple images by iterating through an 'images' list within each data item, processing each image accordingly, and finally attaching these images to the returned dictionary. The implementation assumes an 'images' attribute in the input data, which contains a list of image files to be processed.

**Comment by zhuzilin** [comment]
Thank you so much!
